### PR TITLE
Clean up username generation and document it

### LINKED ISSUE #91: User username field shouldn't get a bytes value
The Django username field should get a unicode value in Python 2 and a string value in Python 3.

However, it's getting the output of `default_username_algo` which generates a string in Python 2 (which is fine since there are no non-ascii characters in it) and a bytes in Python 3 (which is probably not fine).

Further, the signature for username generation functions is "weird" in the sense that it takes a string/bytes as an argument and returns a string/bytes--but it operates on an email address and returns a username. That feels weird.

This issue covers fixing this part of the API.

### PR DESCRIPTION
**Pull request by willkg**
* fix default_username_algo to return the correct types and add tests
* add docs regarding how mozilla-django-oidc generates Django User accounts and
  how to change how usernames are generated

Fixes #91.

### CODE REVIEW

File: docs/installation.rst [hunk 1]
```diff
@@ -121,3 +121,71 @@ email address. Then we could do this:
...
+
+   from mozilla_django_oidc.auth import OIDCAuthenticationBackend
+   from myapp.models import Profile
+
+   class MyOIDCAB(OIDCAuthenticationBackend):
+       def create_user(self, claims):
+           user = super(OIDCAuthenticationRequestView, self).create_user(claims)
+           profile = Profile(user=user, is_new=True)
+
+           return user
```
Comment (willkg) [inline docs/installation.rst]: I don't know if this example is too contrived to be helpful. Is there a better example/use-case we could use here?

Comment (johngian) [inline docs/installation.rst]: I guess introspecting the OIDC authentication scope and getting the profile data might be more relevant, although this one is also useful.

https://openid.net/specs/openid-connect-core-1_0.html#StandardClaims

File: docs/installation.rst [hunk 2]
```diff
@@ -121,3 +121,71 @@ email address. Then we could do this:
 Then you'd use the Python dotted path to that class in the
 ``settings.AUTHENTICATION_BACKENDS`` instead of
 ``mozilla_django_oidc.auth.OIDCAuthenticationBackend``.
+
+
+Creating a Django ``User`` record for new users
+-----------------------------------------------
+
+If a user logs into your site and doesn't already have an account, by default,
```
Comment (akatsoulas) [inline docs/installation.rst]: nit: Shall we also mention something about the `OIDC_CREATE_USER`?  Eg that the default behavior can be changed by the said variable or does it feel too verbose?

Comment (willkg) [inline docs/installation.rst]: It's definitely worth mentioning `OIDC_CREATE_USER`, but I don't really understand it.

If you have it set to False, how does the flow work? Don't you end up with a user that's authenticated, but has no corresponding Django user? What do you do at that point?

File: docs/installation.rst (hunk 1 above)
Comment (willkg) [inline docs/installation.rst]: Ahhh--yeah! That's a much better use-case. I'll change the example.

File: docs/installation.rst (hunk 2 above)
Comment (akatsoulas) [inline docs/installation.rst]: That's a good point. A possible scenario could be the one where the `authenticate` method can be overridden for the use case where a developer needs just to verify an email/user data like eg in a campaign and save them (completely ignoring the Django user). That said, probably this is a hack that does not conform with the Django flow and needs some more work/thought. An idea could be to handle cases like the above at an earlier stage of the flow. @willkg @johngian thoughts?

Comment (willkg) [inline docs/installation.rst]: I like that there's an "escape hatch", but I think any documentation for the setting should probably include some guidance on why you'd want to use it and how to use it.

I think I want to push off thinking about this and any resolution to another issue/PR and land this one as is in regards to `OIDC_CREATE_USER`.

Is that ok? If so, I'll write up an issue about it.

Comment (willkg) [inline docs/installation.rst]: @akatsoulas I wrote up #96 to cover documenting `OIDC_CREATE_USER`.

**Comment by willkg** [comment]
I think we've worked through all the outstanding issues in this PR.

Did I miss anything?

**Pull request merged**